yoyodyne/models/modules/transformer.py

```diff
@@ -184,8 +223,248 @@ def name(self) -> str:
         return "transformer"
 
 
+class TransformerDecoderSeperateFeatures(nn.TransformerDecoder):
+    """A Transformer decoder with seperate features.
+
+    Adding seperate features into the transformer stack is implemented with
+    TransformerDecoderLayerSeperateFeatures layers.
+    """
+
+    def forward(
+        self,
+        target: torch.Tensor,
+        memory: torch.Tensor,
+        features_memory: torch.Tensor,
+        target_mask: Optional[torch.Tensor] = None,
+        memory_mask: Optional[torch.Tensor] = None,
+        features_memory_mask: Optional[torch.Tensor] = None,
+        target_key_padding_mask: Optional[torch.Tensor] = None,
+        memory_key_padding_mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """Pass the inputs (and mask) through the decoder layer in turn.
+
+        Args:
+            target (torch.Tensor): the sequence to the decoder.
+            memory (torch.Tensor): the sequence from the last layer of the
+                encoder.
+            features_memory (torch.Tensor): the sequence from the last layer
+                of the features encoder.
+            target_mask (Optional[torch.Tensor], optional): the mask for the
+                target sequence. Defaults to None.
+            memory_mask (Optional[torch.Tensor], optional): the mask for the
+                memory sequence. Defaults to None.
+            features_memory_mask (Optional[torch.Tensor], optional): the mask
+                for the features. Defaults to None.
+            target_key_padding_mask (Optional[torch.Tensor], optional): the
+                mask for the target keys per batch. Defaults to None.
+            memory_key_padding_mask (Optional[torch.Tensor], optional): the
+                mask for the memory keys per batch. Defaults to None.
+
+        Returns:
+            torch.Tensor: Output tensor.
+        """
+        output = target
+
+        for mod in self.layers:
+            output = mod(
+                output,
+                memory,
+                features_memory,
+                target_mask=target_mask,
+                memory_mask=memory_mask,
+                features_memory_mask=features_memory_mask,
+                target_key_padding_mask=target_key_padding_mask,
+                memory_key_padding_mask=memory_key_padding_mask,
+            )
+        if self.norm is not None:
```

Comment:
possibly dumb question, does this need to be nested inside the for-loop that begins at line 268?

Reply:
Oh i will double check.

Reply:
Adding a norm after each layer looks right.

Reply:
Actually, I am not sure I see this in the original paper, but it is there in the torch implementation.

Reply:
Ah, it is optional and we do not set it. I don't see a citation in their code for when you'd want to set it, but we should keep it here in case that option is set.

Reply:
Okay but maybe we should make a note to try layer norm. I thought that was the standard for transformer encoder layers.